libvast/src/system/posix_file_system.cpp

```diff
@@ -0,0 +1,80 @@
+
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/system/posix_file_system.hpp"
+
+#include "vast/chunk.hpp"
+#include "vast/io/read.hpp"
+#include "vast/io/write.hpp"
+
+#include <caf/config_value.hpp>
+#include <caf/dictionary.hpp>
+#include <caf/result.hpp>
+#include <caf/settings.hpp>
+
+namespace vast::system {
+
+file_system_type::behavior_type posix_file_system(
+  file_system_type::stateful_pointer<posix_file_system_state> self, path root) {
+  return {
+    [=](atom::write, const path& filename,
+        chunk_ptr chk) -> caf::result<atom::ok> {
+      if (auto err = io::write(root / filename, as_bytes(chk))) {
```

Comment:
How does `root / filename` interact with absolute paths? Is it a requirement that `filename` is a relative path? This also applies to the other handlers below.

This should go into the comments for the typed actor behavior in the header, so it's clear what kind of `path` is expected here.

Reply:
Yes, all operations should always be relative to the filesystem root. I'll add documentation.

Reply:
In that case, can you return an error of the path is outside the filesystem root?

Reply:
I think the user should have the assumption that every path in an operation is relative to the filesystem root, but the user shouldn't know where that root is. I've clarified this in the comments.

Reply:
As is this still allows users to escape the filesystem root by specifying ".." as the path. We should guard against that, I think.

Reply:
I don't see the use case. This is not about path sanitization from a malicious standpoint. We'd have to do a lot more to get that right. This would also require going through the code base at many more places beyond this one.

Don't get me wrong, it's not a bad idea, it's just out of scope for this PR.